test/tools/router_check/source/router.cc

```diff
@@ -0,0 +1,213 @@
+
+  if (loader == nullptr) {
+    return false;
+  }
+
+  bool no_failures = true;
+  // Iterate through each test case
+  for (const Json::ObjectPtr& check_config : loader->asObjectArray()) {
+    // Load parameters from json
+    ToolConfig tool_config;
+    tool_config.parseFromJson(check_config);
+
+    Json::ObjectPtr check_type = check_config->getObject("check");
+
+    // Call appropriate function for each match case
+    if (check_type->hasObject("path_redirect")) {
+      if (!compareRedirectPath(tool_config, check_type->getString("path_redirect"))) {
+        no_failures = false;
+      }
+    }
+
+    if (check_type->hasObject("cluster_name")) {
+      if (!compareCluster(tool_config, check_type->getString("cluster_name"))) {
+        no_failures = false;
+      }
+    }
+
+    if (check_type->hasObject("virtual_cluster_name")) {
+      if (!compareVirtualCluster(tool_config, check_type->getString("virtual_cluster_name"))) {
+        no_failures = false;
+      }
+    }
+
+    if (check_type->hasObject("virtual_host_name")) {
+      if (!compareVirtualHost(tool_config, check_type->getString("virtual_host_name"))) {
+        no_failures = false;
+      }
+    }
+
+    if (check_type->hasObject("path_rewrite")) {
+      if (!compareRewritePath(tool_config, check_type->getString("path_rewrite"))) {
+        no_failures = false;
+      }
+    }
+
+    if (check_type->hasObject("host_rewrite")) {
+      if (!compareRewriteHost(tool_config, check_type->getString("host_rewrite"))) {
+        no_failures = false;
+      }
+    }
+  }
+
+  return no_failures;
+}
+
+void ToolConfig::parseFromJson(const Json::ObjectPtr& check_config) {
+  // Extract values from json object
+  bool internal = check_config->getBoolean("internal", false);
+  bool ssl = check_config->getBoolean("ssl", false);
+  std::string authority = check_config->getString("authority");
+  std::string method = check_config->getString("method", "GET");
+  std::string path = check_config->getString("path");
+
+  random_value_ = check_config->getInteger("random_value", 0);
+
+  // Add :method to header
+  headers_.addViaCopy(":method", method);
+
+  // Add ssl header
+  headers_.addViaCopy("x-forwarded-proto", ssl ? "https" : "http");
+
+  // Add authority and path to header
+  headers_.addViaCopy(":authority", authority);
+  headers_.addViaCopy(":path", path);
+
+  // Add internal route status to header
+  if (internal) {
+    headers_.addViaCopy("x-envoy-internal", "true");
+  }
+
+  // Add additional headers
+  if (check_config->hasObject("additional_headers")) {
+    for (const Json::ObjectPtr& header_config :
+         check_config->getObjectArray("additional_headers")) {
+      headers_.addViaCopy(header_config->getString("name"), header_config->getString("value"));
+    }
+  }
+}
+
+bool RouterCheckTool::compareCluster(ToolConfig& tool_config, const std::string expected) {
+  Router::RouteConstSharedPtr route =
```

Comment:
I think the route determination should be done once in `routerCheckTool::compareEntriesInJson`, rather multiple times in each of the `compare*` methods. You can pass a reference to the returned route, so this is just a small refactoring.